synapse/handlers/device.py

```diff
@@ -673,13 +678,135 @@ async def store_dehydrated_device(
             device_id,
             initial_device_display_name,
         )
-        old_device_id = await self.store.store_dehydrated_device(
-            user_id, device_id, device_data
-        )
+
+        time_now = self.clock.time_msec()
+
+        if keys_for_device:
+            keys = await self._check_and_prepare_keys_for_dehydrated_device(
+                user_id, device_id, keys_for_device
+            )
+            old_device_id = await self.store.store_dehydrated_device(
+                user_id, device_id, device_data, time_now, keys
+            )
+        else:
+            old_device_id = await self.store.store_dehydrated_device(
+                user_id, device_id, device_data, time_now
+            )
+
         if old_device_id is not None:
             await self.delete_devices(user_id, [old_device_id])
+
         return device_id
 
+    async def _check_and_prepare_keys_for_dehydrated_device(
+        self, user_id: str, device_id: str, keys: JsonDict
+    ) -> dict:
```

Comment:
Aren't we creating the dehydrated device now? How can it *already* have keys? Aren't devices unique per user/device ID?

Reply:
This is creating/storing the dehydrated device - to your question of how it can already have keys I actually don't know - per the MSC the keys were to be uploaded over the `/keys/upload` endpoint, implying that they already exist: "After the dehydrated device is uploaded, the client will upload the encryption
keys using `POST /keys/upload/{device_id}`, where the `device_id` parameter is
the device ID given in the response to `PUT /dehydrated_device`"

https://github.com/matrix-org/synapse/pull/15929 changed this so that the key upload was integrated into the `PUT` endpoint, and this PR is further refining the key upload such that storing the device and storing the keys are part of one single database transaction, addressing @poljar's concerns around storing the dehdrated device and it's keys being atomic.

Reply:
Uploading the public keys at the time of the dehydrated device creation tries to address this race: https://github.com/matrix-org/matrix-spec-proposals/pull/3814#discussion_r1107879800.

As to how can it already have keys, well if we remember what a dehydrated device is then it becomes quite clear. A dehydrated device are the private identity and one-time keys of a device, of course they are encrypted so the server can't access them.

It becomes quite natural, that we would like to upload the public parts of those same keys to the server at the same time, once we realize this. I think that doing it in a separate `POST /keys/upload/` call is a mistake and quite weird from an API point of view, which the above mentioned race condition showcases.

Reply:
OK, but part of this function checks if we already have keys uploaded for this device (and bails if they're not exactly equal to the new keys). I still don't follow the order of events that would let you upload keys *before* the device is created.

Reply:
This could be me being overzealous here - I basically made sure that all the checks that happen in the `/keys/upload` pathway happen in this pathway, but maybe that's not necessary and we can just store the keys without checking them?

Reply:
> Doing only the checks that make sense in this situation

The only checks done are the ones checking for duplicates but I don't have a strong enough sense of the crypto side of this to know if the keys uploaded for the dehydrated device need to be checked to see if they are duplicates/already stored in the DB - @poljar perhaps you can weigh in on this?

Hopefully they don't and I can just drop the checks because option 2 will require some possibly weird refactoring (which I tried to avoid by committing the cardinal sin of copy/pasting).

Reply:
The way we use the API, they are guaranteed to be unique. Furthermore since we only upload keys once for a dehydrated device there's little chance of getting this wrong.

A slight digression, I think that the server should not refuse duplicate keys in any case, even for the normal `/keys/upload` endpoint, it should only notify the client that they are uploading duplicates. But that's topic for another MSC.

Reply:
Part of it would be straightforward (Move `_set_e2e_device_keys_txn` to a method (instead of an inner function) and call it from both places.)

Although it seems this entire function is just a copy of `upload_keys_for_user` -- can we just call that instead?

Reply:
> The way we use the API, they are guaranteed to be unique. Furthermore since we only upload keys once for a dehydrated device there's little chance of getting this wrong.

I think if this is the case we can probably call the store methods that set the values directly and avoid the complicated logic of reusing keys.

Reply:
They would need to be refactored to take a transaction as the first parameter and current callers would call that also.

Reply:
Right I think this is sorted now, sorry for the confusion.